### PR TITLE
Vault Agent Template

### PR DESCRIPTION
**Pull request by catsby**
This pull request adds a package `template` to Vault Agent. The new type `template.Server` (aka "TemplateServer") accepts Consul Template style template configurations, and then constructs and manages the lifecycle of a Consul Template Runner that will communicate with a Vault server. 

The TemplateServer operates in the same manner as `SinkServer`, receiving a token from `AuthHandler`. When a new token is received the TemplateServer updates it's internal Vault client with the token updates the internal runner. The runner is responsible for querying the Vault server and rendering the template. 

The TemplateServer needs to manage the lifecycle of the runner and report any errors encountered. 

There are caveats:

- Limited test coverage. More tests will come in future commits and/or separate pull requests 
- No configuration hot-reloads. Originally planned to include this but found that Vault Agent does not support this at all yet, so it will require more engineering affecting more of Agent rather than just reloading the templates. 
- Consul Template does not provide a way to supply a pre-configured Vault client. Originally had planned to just share the client that's already configured and used by `AuthHandler` but there's no easy way to do this. As a result, the `setupTemplateConfig` method was added to read Vault configuration values from the command, which is already setup with values from the environment. Those values are then passed on to Consul Template which builds it's own internal Vault client.

### CODE REVIEW

File: command/agent.go [hunk 1]
```diff
@@ -507,21 +534,30 @@ func (c *AgentCommand) Run(args []string) int {
...
+	case <-ctx.Done():
+	case <-unblockCh:
+	}
+
 	select {
 	case <-ssDoneCh:
 		// This will happen if we exit-on-auth
 		c.logger.Info("sinks finished, exiting")
+	// case <-c.SighupCh:
+	// TODO: implement reloading
```
Comment (calvn) [inline command/agent.go]: Should this be removed since hot-reloading is not supported yet?

File: command/agent.go [hunk 2]
```diff
@@ -614,3 +654,34 @@ func (c *AgentCommand) removePidFile(pidPath string) error {
...
 }
+
+// setupTemplateConfig creates a config.Vault struct for use by Consul Template.
+// Consul Template does not currently allow us to pass in a configured API
+// client, unlike the AuthHandler and SinkServer that reuse the client creted in
+// this Run() method. Here we build a config.Vault struct for use by the
+// Template Server that matches the configuration used to create the client
+// (c.client), but in a struct of type config.Vault so that Consul Template can
+// create it's own api client internally.
+// TODO test setupTemplateConfig
```
Comment (calvn) [inline command/agent.go]: This todo line can be removed

File: command/agent/template/template.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+						Token: latestToken,
+					},
+				}
+				runnerConfig.Merge(&ctv)
+				runnerConfig.Finalize()
+				var runnerErr error
+				ts.runner, runnerErr = manager.NewRunner(runnerConfig, false)
+				if runnerErr != nil {
+					ts.logger.Error("template server failed with new Vault token", "error", runnerErr)
+					// TODO: continue or fail here? Right now we just continue
```
Comment (calvn) [inline command/agent/template/template.go]: I think it's fine to continue in here and wait for the next incoming token.

File: command/agent/template/template_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	if ch == nil {
+		t.Fatal("nil blocking channel returned")
+	}
+}
+
+func TestServerRun(t *testing.T) {
+	// create http test server
+	ts := httptest.NewServer(http.HandlerFunc(handleRequest))
+	defer ts.Close()
+	tmpDir, err := ioutil.TempDir("", "agent-tests")
```
Comment (calvn) [inline command/agent/template/template_test.go]: You'll need to clean the dir up with `defer os.RemoveAll(tmpDir)`

File: command/agent/template/template_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	type secretRender struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+		Version  string `json:"version"`
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			for i, template := range tc.templates {
+				// template.Destination = testhelpers.StrPtr(fmt.Sprintf("%s/render_%d.txt", tmpDir, i))
```
Comment (calvn) [inline command/agent/template/template_test.go]: Commented block here.

File: command/agent.go [hunk 6]
```diff
@@ -614,3 +654,34 @@ func (c *AgentCommand) removePidFile(pidPath string) error {
...
+func (c *AgentCommand) setupTemplateConfig() *config.Vault {
+	cfg := new(config.Vault)
```
Comment (kalafut) [inline command/agent.go]: Noted previously but I didn't see a resolution. Can't this initialization be simplified to:
```
return &config.Vault {
    Address:   c.flagAddress,
    CACert:    c.flagCACert,
    ...
}
```

File: command/agent/template/template.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	conf.Vault.Address = &sc.VaultConf.Address
+
+	if sc.Namespace != "" {
+		conf.Vault.Namespace = &sc.Namespace
+	}
+
+	conf.Vault.SSL = &ctconfig.SSLConfig{
+		Enabled:    boolPtr(false),
+		Verify:     boolPtr(false),
+		Cert:       &emptyStr,
```
Comment (kalafut) [inline command/agent/template/template.go]: Mentioned on the other PR and copying here: A `strPtr()` function seems useful both for consistency and to break the pattern where all of these are being assigned the same pointer. e.g. if somewhere later I do `*conf.Vault.Token = "abc"`, then _all_ of the `&emptyStr` config items would be suddenly be "abc". Breaking the aliasing between these parameters is desirable.

File: command/agent.go (hunk 1 above)
Comment (catsby) [reply command/agent.go]: Removed in 2ac3145e5

File: command/agent.go (hunk 2 above)
Comment (catsby) [reply command/agent.go]: Removed in 2ac3145e5

File: command/agent.go [hunk 8]
```diff
@@ -614,3 +654,34 @@ func (c *AgentCommand) removePidFile(pidPath string) error {
 	}
 	return os.Remove(pidPath)
 }
+
+// setupTemplateConfig creates a config.Vault struct for use by Consul Template.
+// Consul Template does not currently allow us to pass in a configured API
+// client, unlike the AuthHandler and SinkServer that reuse the client creted in
```
[the remark this answers is not in the document]
Comment (catsby) [reply command/agent.go]: Corrected in 2ac3145e5

File: command/agent.go (hunk 6 above)
Comment (catsby) [reply command/agent.go]: Updated in 2ac3145e5

File: command/agent/template/template.go (hunk 3 above)
Comment (catsby) [reply command/agent/template/template.go]: 👍 I removed the comment in 70e505d7c

File: command/agent/template/template_test.go (hunk 5 above)
Comment (catsby) [reply command/agent/template/template_test.go]: Cleaned up in 70e505d7c, thanks

File: command/agent/template/template_test.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+		"basic": {
+			templates: []*ctconfig.TemplateConfig{
+				&ctconfig.TemplateConfig{
+					Contents: testhelpers.StrPtr(templateContents),
+				},
+			},
+		},
+	}
+
+	templateTokenCh := make(chan string, 1)
```
[the remark this answers is not in the document]
Comment (catsby) [reply command/agent/template/template_test.go]: Good catch, thanks. Fixed in 2c59414f7

File: command/agent/template/template_test.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+      "version": 3
+    }
+  },
+  "wrap_info": null,
+  "warnings": null,
+  "auth": null
+}
+`
+
+var templateContents = `
```
[the remark this answers is not in the document]
Comment (catsby) [reply command/agent/template/template_test.go]: I anticipate this simple template being reused across other tests, so I'd like to leave it here for now. Are there significant drawbacks to sharing inputs across tests that I should be aware of, in the context of rendering this simple template?

File: command/agent/template/template_test.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+					t.Fatal("nil blocking channel returned")
+				}
+
+				go server.Run(ctx, templateTokenCh, tc.templates)
+
+				// send a dummy value to trigger the internal Runner to query for secret
+				// info
+				templateTokenCh <- "test"
+
+				// Unblock should close immediately b/c there are no templates to render
```
[the remark this answers is not in the document]
Comment (catsby) [reply command/agent/template/template_test.go]: Old, outdated comment, removed in 2c59414f7 , thanks!

File: command/agent/template/template.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+// the Runner and return
+func (ts *Server) Run(ctx context.Context, incoming chan string, templates []*ctconfig.TemplateConfig) {
+	latestToken := new(string)
+	ts.logger.Info("starting template server")
+	defer func() {
+		ts.logger.Info("template server stopped")
+		close(ts.DoneCh)
+	}()
+
+	if incoming == nil {
```
[the remark this answers is not in the document]
Comment (ncabatoff) [reply command/agent/template/template.go]: I guess it's fine for now, I don't to delay this PR over something that can't happen with the existing prod code.  We can always revisit later.

File: command/agent/template/template_test.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+				select {
+				case <-ctx.Done():
+				case <-unblockCh:
+				}
+
+				// cancel to clean things up
+				cancelFunc()
```
[the remark this answers is not in the document]
Comment (catsby) [reply command/agent/template/template_test.go]: I didnt defer because I wanted the temporary directory cleaned up. Defer would just cause all that to pile up until the entire test finished, correct?

[the remark this answers is not in the document]
Comment (catsby) [reply command/agent/template/template_test.go]: Aside, this made me realize I had structured this test poorly; I was testing the rendering of all templates in a loop of each template. If the test had multiple templates this probably would have been noticeable, but it only has 1 at the moment. I updated that in 05a85fea6 , and it will need further updates (the value checking is hard coded to this single template), but that can come in a follow-up PR if you're not opposed to it

File: command/agent/template/template_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply command/agent/template/template_test.go]: It's definitely an opinion I have which is not universal across the team.  But I've seen plenty of test suites in my life that followed that pattern, which over the years became really hard to follow as a result.  The shared input tends to be overcomplicated in order to cater to each test's needs, and the problem only gets worse with time.   You can't really tell without great effort what each test depends on in their shared input.   Because the shared input is liable to contain things not needed by a given test, it reduces the clarity of what the test is actually doing.  And because it's unclear just what's needed for a minimal test, developers writing new tests are incentivized to double down and further exacerbate the problem.

File: command/agent/template/template.go (hunk 7 above)
Comment (catsby) [reply command/agent/template/template.go]: Updated in 778721e19 and ca0bb4d1f

File: command/agent.go [hunk 14]
```diff
@@ -536,6 +563,12 @@ func (c *AgentCommand) Run(args []string) int {
 		}
 	}()
 
+	// Wait for the template to render
+	select {
+	case <-ctx.Done():
+	case <-ts.UnblockCh:
```
Comment (calvn) [inline command/agent.go]: Would this block forever if `ts` is not initialized?

Comment (catsby) [reply command/agent.go]: It would. I suppose that would happen if no auth-auth / method was setup. Fixed in 419f21b95

File: command/agent/template/template.go [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+			ts.logger.Error("template server error", "error", err.Error())
+			close(ts.UnblockCh)
+			return
+		case <-ts.runner.TemplateRenderedCh():
+			// A template has been rendered, unblock
+			if ts.exitAfterAuth {
+				// if we want to exit after auth, go ahead and shut down the runner
+				ts.runner.Stop()
+			}
+			close(ts.UnblockCh)
```
[the remark this answers is not in the document]
Comment (calvn) [reply command/agent/template/template.go]: Should we also `return` after this?
